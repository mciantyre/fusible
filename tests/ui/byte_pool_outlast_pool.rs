// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

use fusible::byte_pool::{BytePool, Bytes};

/// A smart pointer cannot outlive the pool that derives it.
fn main() {
    let mut storage = [0u8; 1024];
    let floats: Bytes<[f32]> = {
        let pool = core::pin::pin!(BytePool::context());
        let pool = BytePool::create(pool.into_ref(), &mut storage, &Default::default()).unwrap();
        let floats = pool.allocate(|| [0f32, 1f32, 2f32]).unwrap();
        Bytes::into_slice(floats)
    };
    drop(floats);
}
