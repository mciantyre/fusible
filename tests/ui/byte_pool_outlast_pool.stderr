error[E0716]: temporary value dropped while borrowed
  --> tests/ui/byte_pool_outlast_pool.rs:10:20
   |
9  |     let floats: Bytes<[f32]> = {
   |         ------ borrow later stored here
10 |         let pool = core::pin::pin!(BytePool::context());
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
...
14 |     };
   |     - temporary value is freed at the end of this statement
   |
   = note: consider using a `let` binding to create a longer lived value
   = note: this error originates in the macro `core::pin::pin` (in Nightly builds, run with -Z macro-backtrace for more info)
